### PR TITLE
Store the peer list on the DB of a node

### LINKED ISSUE #1569: Save the peer list on the disk
To reduce bootstrap nodes' burden, we need to save the received peer list in the disk.

### PR DESCRIPTION
**Pull request by MSNTCS**
Storing all stable nodes that have been connected to the node in the database. It fixes #1569".

### CODE REVIEW

File: rpc/src/v1/impls/mempool.rs [hunk 1]
```diff
@@ -97,7 +97,6 @@ where
         }
     }
 
-
```
Comment (sgkim126) [inline rpc/src/v1/impls/mempool.rs]: Please remove these kindes of changes not related to this commit.

File: network/src/p2p/mod.rs [hunk 2]
```diff
@@ -15,7 +15,7 @@
 // along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
 mod connection;
-mod handler;
+pub mod handler;
```
Comment (sgkim126) [inline network/src/p2p/mod.rs]: IMO, Reexporting `ManagingPeerdb` would be better than making the `handler` module public.

File: core/src/consensus/tendermint/engine.rs [hunk 3]
```diff
@@ -314,11 +313,13 @@ impl ConsensusEngine for Tendermint {
         let client = self.client().ok_or(EngineError::CannotOpenBlock)?;
         let block_hash = match block_number {
             None => {
-                client.block_header(&BlockId::Latest).expect("latest block must exist").hash() // the latest block
+                client.block_header(&BlockId::Latest).expect("latest block must exist").hash()
+                // the latest block
```
Comment (sgkim126) [inline core/src/consensus/tendermint/engine.rs]: I think this comment is not related to this PR.

File: core/src/consensus/tendermint/engine.rs [hunk 4]
```diff
@@ -314,11 +313,13 @@ impl ConsensusEngine for Tendermint {
...
             }
             Some(block_number) => {
                 assert_ne!(0, block_number);
-                client.block_header(&(block_number - 1).into()).ok_or(EngineError::CannotOpenBlock)?.hash() // the parent of the given block number
+                client.block_header(&(block_number - 1).into()).ok_or(EngineError::CannotOpenBlock)?.hash()
+                // the parent of the given block number
```
Comment (sgkim126) [inline core/src/consensus/tendermint/engine.rs]: ditto.

File: core/src/db.rs [hunk 5]
```diff
@@ -37,6 +37,8 @@ pub const COL_MEMPOOL: Option<u32> = Some(4);
 pub const COL_ERROR_HINT: Option<u32> = Some(5);
 /// Number of columns in DB
 pub const NUM_COLUMNS: Option<u32> = Some(6);
+/// Column for Storing Peer list
+pub const COL_PEER: Option<u32> = Some(7);
```
Comment (sgkim126) [inline core/src/db.rs]: It should be six and you should increase `NUM_COLUMNS` to seven.

File: core/src/peer_db.rs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+use crate::db::COL_PEER;
+use cnetwork::ManagingPeerdb;
+use kvdb::{DBTransaction, KeyValueDB};
+use std::sync::Arc;
+
+const COLUMN_TO_WRITE: Option<u32> = COL_PEER;
```
Comment (sgkim126) [inline core/src/peer_db.rs]: :+1:

File: core/src/peer_db.rs [hunk 7]
```diff
@@ -0,0 +1,44 @@
+// Copyright 2018-2019 Kodebox, Inc.
```
Comment (sgkim126) [inline core/src/peer_db.rs]:
```
// Copyright 2020 Kodebox, Inc.
```

File: network/src/p2p/handler.rs [hunk 8]
```diff
@@ -1129,15 +1141,25 @@ impl IoHandler<Message> for Handler {
...
                     }
                 } else {
                     cdebug!(NETWORK, "Invalid outgoing token({}) on deregister", stream);
+                    None
                 }
             }
             _ => unreachable!("Unexpected stream on deregister: {}", stream),
+        };
+
+        if let Some(_address) = peer_socket_adress {
```
Comment (sgkim126) [inline network/src/p2p/handler.rs]: I don't think using an Option for this is a good indication of intent.
How about making this code a function instead of using `peer_socket_adress`?

Comment (MSNTCS) [reply network/src/p2p/handler.rs]: Can you provide more details? Do you think it is better to put a function in the first and second match?

Comment (sgkim126) [reply network/src/p2p/handler.rs]:
```rust
            let batch = DBTransaction::new();
            let remove_target = peer_socket_adress;
            let s = rlp::encode(&remove_target);
            self.peer_db.delete(batch, &s);
```
I meant extracting the code above into a method and calling the method at line 1065 and 1084.

**Comment by majecty** [comment]
Please add an issue number in the description of PR like "Fixes #1569".
Add a description in the commit message and PR description about why we need to save the peer list in the DB and what will be changed after this PR or commit.

Please use a full sentence in the description of PR and commit message.

File: network/src/p2p/connection/established.rs [hunk 9]
```diff
@@ -33,7 +33,7 @@ use crate::SocketAddr;
 
 pub struct EstablishedConnection {
     stream: SignedStream,
-    peer_addr: SocketAddr,
+    pub peer_addr: SocketAddr,
```
Comment (sgkim126) [inline network/src/p2p/connection/established.rs]: Why do you make it public?

File: network/src/p2p/handler.rs [hunk 10]
```diff
@@ -493,6 +491,13 @@ impl IoHandler<Message> for Handler {
                 is_inbound: true,
             } => {
                 let mut inbound_connections = self.inbound_connections.write();
+                let batch = DBTransaction::new();
+                let target = connection.peer_addr();
+                let s = rlp::encode(target);
+                let time: u64 =
+                    SystemTime::now().duration_since(UNIX_EPOCH).expect("There is no time machine.").as_secs();
+                let key = rlp::encode(&time);
+                self.peer_db.insert(batch, &s, &key);
```
Comment (sgkim126) [inline network/src/p2p/handler.rs]: It seems that `key` is wrongly named.

File: network/src/p2p/handler.rs [hunk 11]
```diff
@@ -1057,6 +1062,10 @@ impl IoHandler<Message> for Handler {
                     self.routing_table.remove(con.peer_addr());
                     self.inbound_tokens.lock().restore(stream);
                     ctrace!(NETWORK, "Inbound connect({}) removed", stream);
+                    let batch = DBTransaction::new();
+                    let remove_target = con.peer_addr();
+                    let s = rlp::encode(remove_target);
+                    self.peer_db.delete(batch, &s);
```
Comment (sgkim126) [inline network/src/p2p/handler.rs]: You missed removing an address of the outbound connection at line 1085.

File: network/src/lib.rs [hunk 12]
```diff
@@ -66,6 +67,8 @@ pub use crate::service::{Error as NetworkServiceError, Service as NetworkService
 
 pub use crate::filters::{FilterEntry, Filters, FiltersControl};
 pub use crate::routing_table::RoutingTable;
+pub use p2p::Handler;
```
Comment (sgkim126) [inline network/src/lib.rs]:
```rust
pub use self::p2p::{Handler, ManagingPeerdb};
```

File: network/src/p2p/handler.rs [hunk 13]
```diff
@@ -16,9 +16,10 @@
 
 use std::collections::{BTreeSet, HashMap, VecDeque};
 use std::iter::FromIterator;
+use std::marker::{Send, Sync};
```
Comment (sgkim126) [inline network/src/p2p/handler.rs]: You don't need these imports.
https://doc.rust-lang.org/std/prelude/index.html

File: network/src/p2p/handler.rs [hunk 14]
```diff
@@ -80,14 +82,19 @@ const RETRY_SYNC_MAX: Duration = Duration::from_secs(10); // T1
 const RTT: Duration = Duration::from_secs(10); // T2
 const WAIT_SYNC: Duration = Duration::from_secs(30); // T3 >> T1 + RTT
 
+pub trait ManagingPeerdb: Send + Sync {
+    fn insert(&self, transaction: DBTransaction, key: &[u8], value: &[u8]);
```
Comment (sgkim126) [inline network/src/p2p/handler.rs]: Does `ManagingPeerdb` need `DBTransaction`?
The implementation might use a transaction but it seems implementation detail.
I think there is no need to pass the transaction.

Comment (sgkim126) [inline network/src/p2p/handler.rs]: The arguments are too generic.
Do you have a plan to insert other types except `SocketAddr`?
If not, I think the key type should be `SocketAddr` or `&SocketAddr`.

Comment (sgkim126) [reply network/src/p2p/handler.rs]: And what does the inserting value mean?
IMO, it is an implementation detail not for an interface.

Comment (MSNTCS) [reply network/src/p2p/handler.rs]: I cannot understand the second one. Can you explain more? 
I mean this one 
https://github.com/CodeChain-io/codechain/pull/1945/commits/916565582681895600fe754c43a0a078f3011e9c#r363035953

Comment (sgkim126) [reply network/src/p2p/handler.rs]: Currently, the `insert` method takes an rlp encoded socket address as `key` and an rlp encoded time as `value`. The key is necessary; the purpose of this trait is storing the address. I think the purpose of the time is to know how the connection is stable, but I wonder whether it's the role of `ManagingPeerdb`. How to store the address and what address to use are implementation details.

File: network/src/p2p/handler.rs (hunk 8 above)
Comment (MSNTCS) [reply network/src/p2p/handler.rs]: Done

File: network/src/p2p/handler.rs (hunk 10 above)
Comment (MSNTCS) [reply network/src/p2p/handler.rs]: Done

File: network/src/lib.rs (hunk 12 above)
Comment (MSNTCS) [reply network/src/lib.rs]: Done

File: network/src/p2p/handler.rs (hunk 13 above)
Comment (MSNTCS) [reply network/src/p2p/handler.rs]: Done

File: network/src/p2p/handler.rs (hunk 14 above)
Comment (MSNTCS) [reply network/src/p2p/handler.rs]: Hopefully done.

Comment (MSNTCS) [reply network/src/p2p/handler.rs]: Got it and apply you desire changes.

**Comment by MSNTCS** [comment]
I tested the unpassed test( `Change common Params that doesn't affect validator set`) on my local machine and it passed the test. My implementation does not have any problem with this test.

File: core/src/db.rs [hunk 15]
```diff
@@ -35,8 +35,11 @@ pub const COL_EXTRA: Option<u32> = Some(3);
 pub const COL_MEMPOOL: Option<u32> = Some(4);
 /// Column for Transaction error hints
 pub const COL_ERROR_HINT: Option<u32> = Some(5);
+/// Column for Storing Peer list
+pub const COL_PEER: Option<u32> = Some(6);
 /// Number of columns in DB
-pub const NUM_COLUMNS: Option<u32> = Some(6);
+pub const NUM_COLUMNS: Option<u32> = Some(7);
```
Comment (sgkim126) [inline core/src/db.rs]: @majecty I'm not sure whether it's safe to change the number of columns. Would you check whether it's compatible with the current db?

Comment (majecty) [reply core/src/db.rs]: CodeChain failed to run if the NUM_COLUMNS changed. We might need a migration tool.

Here is the log message:
```
#550732000 2020-01-06 16:35:48.636848000  main WARN kvdb_rocksdb  DB corrupted: Invalid argument: You have to open all column families. Column families not opened: col3, col4, col1, col2, col5, col0, attempting repair
Error: "Low level database error. Some issue with disk?"
```

Comment (majecty) [reply core/src/db.rs]: @sgkim126 What do you think about adding another issue about the migration? Since to handle it the programmer needs to know some RocksDB knowledge, it is not a good first issue.

Comment (sgkim126) [reply core/src/db.rs]: Thank you for checking it @majecty.
@MSNTCS Would you upload this PR to https://github.com/CodeChain-io/foundry first?

Comment (MSNTCS) [reply core/src/db.rs]: @sgkim126 I just did. You can check on the foundry.

Comment (sgkim126) [reply core/src/db.rs]: Hi @MSNTCS,
I have a gut feeling that we can reuse an [extra column](https://github.com/CodeChain-io/codechain/blob/22fe8037f252b21b2db9e92ce747f1bcc39fba7f/core/src/db.rs#L33) for it because `KeyValueDB` has an [interface to iterate the column that has a specific prefix](https://docs.rs/kvdb/0.1.1/kvdb/trait.KeyValueDB.html#tymethod.iter_from_prefix).
This approach can make it difficult to manage later because it differs from the foundry implementation. But it seems to be easier to maintain than adding migration code.
Would you try it?

Comment (MSNTCS) [reply core/src/db.rs]: Hello, @sgkim126, 
I will try to implement it, should it be as another commit or do I need to modify the previous one?

Comment (sgkim126) [reply core/src/db.rs]: I prefer to modify it, but it shouldn't be matter anyway. Please take the way that is comfortable for you.

File: codechain/auto_self_nominate.rs [hunk 16]
```diff
@@ -0,0 +1,156 @@
+// Copyright 2020 Kodebox, Inc.
```
Comment (sgkim126) [inline codechain/auto_self_nominate.rs]: It seems that this commit should not be included here.

**Comment by sgkim126** [comment]
@MSNTCS Would you check the commits included in this PR again?

**Comment by MSNTCS** [comment]
> @MSNTCS Would you check the commits included in this PR again?

Sorry, for the late response. I had changed when I saw your comment. Now is only related commit.

File: core/src/db.rs (hunk 15 above)
Comment (majecty) [reply core/src/db.rs]: This PR is still using COL_PEER.
@MSNTCS Did you try the extra column?

**Comment by sgkim126** [comment]
https://github.com/CodeChain-io/foundry/pull/133 is a suggestion to implement PeerDb without implementing the migration code.

**Comment by MSNTCS** [comment]
@majecty Should we update this based on the new implementation? or can we close this?

**Comment by majecty** [comment]
Let's cherry-pick the Seulgi's commits to this PR
@sgkim126 I want to merge this PR with your commits!

**Comment by MSNTCS** [comment]
> Let's cherry-pick the Seulgi's commits to this PR
> @sgkim126 I want to merge this PR with your commits!

@sgkim126 would you mind answering this?

**Comment by sgkim126** [comment]
@MSNTCS I agree on @majecty. It would be better to merge the commits into this PR.

**Comment by MSNTCS** [comment]
> @MSNTCS I agree on @majecty. It would be better to merge the commits into this PR.

Thanks

**Comment by majecty** [comment]
Please run `rustfmt` whenever you create a commit.

**Comment by majecty** [comment]
Your commit messages "Update handler.rs" and "Update service.rs" does not give any meaningful information to a reader.
It seems that the reason why you created the two commits is that some code is different from Foundry's and yours. Then could you give the information in the commit message?

**Comment by majecty** [comment]
@MSNTCS Could you also cherry pick these commits?
IMO, you can squash them in your first commit.
https://github.com/CodeChain-io/foundry/commit/e22ea52a8a2500db5eda8989d64fddad47949359
https://github.com/CodeChain-io/foundry/commit/c6dcf5421d9e6f186a8eb1c3931bb7e54a859321
https://github.com/CodeChain-io/foundry/commit/82716dd2f6e99fecd8e770fadc530e1c1aa83336
https://github.com/CodeChain-io/foundry/commit/7471a905d139bf575c6735eb94c67fd36e1971e5
https://github.com/CodeChain-io/foundry/commit/d5306453aa25e00f37bbdc11544fc0166eaf7eeb
https://github.com/CodeChain-io/foundry/commit/97b8c62cdce33a292c4dc88cb79dc1e3d01f067f

File: core/src/db.rs (hunk 15 above)
Comment (majecty) [reply core/src/db.rs]: We will import Seulgi's commits from Foundry. The commits make the code to use COL_EXTRA. Let's finish this thread.

File: core/src/db.rs [hunk 17]
```diff
@@ -35,8 +35,11 @@ pub const COL_EXTRA: Option<u32> = Some(3);
...
+
```
Comment (majecty) [inline core/src/db.rs]: @MSNTCS Please remove unrelated new line.

File: network/src/service.rs [hunk 18]
```diff
@@ -18,7 +18,8 @@ use crate::client::Client;
 use crate::control::{Control, Error as ControlError};
 use crate::filters::{FilterEntry, FiltersControl};
 use crate::routing_table::RoutingTable;
-use crate::{p2p, Api, NetworkExtension, SocketAddr};
+use crate::{p2p, Api, ManagingPeerdb, NetworkExtension, SocketAddr};
+
```
Comment (majecty) [inline network/src/service.rs]: Please remove this empty line.

File: network/src/service.rs [hunk 19]
```diff
@@ -28,6 +29,7 @@ use std::collections::HashMap;
 use std::net::IpAddr;
 use std::sync::Arc;
 
+
```
Comment (majecty) [inline network/src/service.rs]: Please remove this empty line.
